File: Python/Programmars/Level_2/괄호 회전하기.py
from collections import deque


def isRight(string):
    stack = []
    start = ['(', '[', '{']
    end = [')', ']', '}']
    for s in string:
        if s in start:
            stack.append(s)
        else:
            if not stack:
                return False
            for i in range(3):
                if s == end[i] and stack[-1] == start[i]:
                    stack.pop()
    if stack:
        return False
    return True


def solution(s):
    answer = 0
    for _ in range(len(s)-1):
        if isRight(s):
            answer += 1
        s = s[1:]+s[0]
    return answer


print(solution("[](){}"))  # 3
print(solution("}]()[{"))  # 2
print(solution("[)(]"))  # 0
print(solution("}}}"))  # 0

File: JavaScript/Programmars/Level_2/괄호 회전하기.js
// 문제: https://programmers.co.kr/learn/courses/30/lessons/76502
const solution = (s) => {
  const parenthesisObj = {
    '(': ')',
    '[': ']',
    '{': '}',
  };

  const isRightString = (str) => {
    const stack = [];
    let next;
    for (let s of str) {
      if (s === '(' || s === '[' || s === '{') {
        stack.push(s);
      } else {
        if (stack.length === 0) return false;
        next = stack.pop();
        if (s !== parenthesisObj[next]) return false;
      }
    }
    return stack.length === 0;
  };

  const sLen = s.length;
  if (sLen % 2 !== 0) return 0;

  for (let [open, close] of Object.entries(parenthesisObj)) {
    if (s.split(open).length !== s.split(close).length) return 0;
  }

  let answer = 0;
  let str = s;
  for (let x = 0; x < sLen; x++) {
    if (isRightString(str)) answer++;
    str = str.substr(1) + str[0];
  }

  return answer;
};

console.log(solution('[](){}')); // 3
console.log(solution('}]()[{')); // 2
console.log(solution('[)(]')); // 0
console.log(solution('}}}')); // 0
